### ISSUE TITLE
Automatic parameter tuning and grid search

### ISSUE THREAD

**Issue by tobigithub** [Microsoft/LightGBM#127]
LightGBM is so amazingly fast it would be important to implement
a native grid search for the single executable EXE that covers the
most common influential parameters such as num_leaves, bins, feature_fraction,
bagging_fraction, min_data_in_leaf, min_sum_hessian_in_leaf and few others.

As simple option for the LightGBM executable auto grid could be 
a simple text based option in the train.conf such as _bagging_freq = [10, 20, 40, 80, 160]_

The output at the end of the run should contain the parameters selected and the
metric for each of the parameters as table. 

That includes also grid searches for the R and existing Python interface. 
Plus, the WIKI examples of the grid search make life much easier for quick adaption.

**Comment by wxchan** [comment]
Can you try [sklearn.model_selection.GridSearchCV](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV) with LightGBM sklearn interface to see if it works? The sklearn interface is inherited from sklearn.base.BaseEstimator so it's expected to work exactly like other estimators in sklearn.

**Comment by wxchan** [comment]
I tested sklearn.model_selection.GridSearchCV, it should work well. Follow #128 .

**Comment by tobigithub** [comment]
@wxchan 
Nice. Thanks.

**Comment by tobigithub** [comment]
oh no.
so what about Microsoft EXE (MZ Mark Zbikowski) file users?
The need to switch to Python?

**Comment by Laurae2** [comment]
@guolinke @tobigithub I think this feature should be handed to the specialized interfaces which are doing hyperparameter tuning and grid searching and not LightGBM itself, unless there is a guaranteed way to get the best parameters specifically for LightGBM only.

If you are an EXE file user, what about a script:

* Creating dynamically configuration files with the appropriate hyperparameters (in .sh/.bat not much a problem for instance)
* Redirect/Capture the printed logs to get the best performance of each trained model
* At the end: compare to previous stored best performance to choose the best set of hyperparameters to use

This should fix your issue for grid searching. For hyperparameter tuning, you should implement your own because there are too many ways to do it.

For R, making a grid search just requires storing best performance in a loop where you used `expand.grid` to get the hyperparameter matrix from an initial matrix. For hyperparameter tuning, this should be handed to specialized optimizers (cross-entropy optimization, bayesian optimization, etc.).

**Comment by tobigithub** [comment]
Thank you!
Tobias

**Comment by hack-r** [comment]
It would be nice if we also had grid search in R. If you created those scripts for the .exe version maybe if you give a link to them it would help us write something similar for R.

**Comment by Laurae2** [comment]
@hack-r this is simple to setup in R:

```r
library(data.table)
library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label, free_raw_data = FALSE)
data(agaricus.test, package = "lightgbm")
test <- agaricus.test
dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)
valids <- list(test = dtest)

grid_search <- expand.grid(Depth = 8,
                           L1 = 0:5,
                           L2 = 0:5)

model <- list()
perf <- numeric(nrow(grid_search))

for (i in 1:nrow(grid_search)) {
  model[[i]] <- lgb.train(list(objective = "regression",
                          metric = "l2",
                          lambda_l1 = grid_search[i, "L1"],
                          lambda_l2 = grid_search[i, "L2"],
                          max_depth = grid_search[i, "Depth"]),
                     dtrain,
                     2,
                     valids,
                     min_data = 1,
                     learning_rate = 1,
                     early_stopping_rounds = 5)
  perf[i] <- min(rbindlist(model[[i]]$record_evals$test$l2))
}
```

Result:

```r
> cat("Model ", which.min(perf), " is lowest loss: ", min(perf), sep = "")
Model 1 is lowest loss: 1.972152e-31
> print(grid_search[which.min(perf), ])
  Depth L1 L2
1     8  0  0
```

**Comment by hack-r** [comment]
@Laurae2 Awesooooome! Thanks!!

**Comment by tobigithub** [comment]
Also this extended code for grid search in R using LightGBM, the only problem it eats a lot of memory, also only uses one CPU core.
```
# https://github.com/Microsoft/LightGBM/blob/ca30b8cb16e535ca5482c127986842bd4d5a99b8/docs/Parameters.md
# Original: Laurae2
# https://github.com/Microsoft/LightGBM/issues/127#issuecomment-267755141

library(data.table)
library(lightgbm)

data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label, free_raw_data = FALSE)

data(agaricus.test, package = "lightgbm")
test <- agaricus.test
dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)

valids <- list(test = dtest)

grid_search <- expand.grid(max_depth = c(10,20,40,80),
                           min_data_in_leaf = c(1,2,4),
                           min_sum_hessian_in_leaf = c(0.05,0.1,0.2),
                           feature_fraction = c(0.8,0.9,0.95),
                           bagging_fraction = c(0.4,0.6),
                           bagging_freq = c(2,4),
                           lambda_l1 = c(0.2,0.4),
                           lambda_l2 = c(0.2,0.4),
                           min_gain_to_split = c(0.2,0.4))

model <- list()
perf <- numeric(nrow(grid_search))

for (i in 1:nrow(grid_search)) {
  model[[i]] <- lgb.train(list(objective = "regression",
                          metric = "l2",
                          max_depth = grid_search[i, "max_depth"],
                          min_data_in_leaf = grid_search[i,"min_data_in_leaf"],
                          min_sum_hessian_in_leaf = grid_search[i, "min_sum_hessian_in_leaf"],
                          feature_fraction =  grid_search[i, "feature_fraction"],
                          bagging_fraction =  grid_search[i, "bagging_fraction"],
                          bagging_freq =  grid_search[i, "bagging_freq"],
                          lambda_l1 =  grid_search[i, "lambda_l1"],
                          lambda_l2 =  grid_search[i, "lambda_l2"],
                          min_gain_to_split =  grid_search[i, "min_gain_to_split"]),
                     dtrain,
                     2,
                     valids,
                     #min_data = 1,
                     num_leaves = 100,
                     learning_rate = 0.1,
                     early_stopping_rounds = 20)
  perf[i] <- min(rbindlist(model[[i]]$record_evals$test$l2))
}
# grid_search
cat("Model ", which.min(perf), " is lowest loss: ", min(perf), sep = "","\n")
print(grid_search[which.min(perf), ])

```
output:

```
> cat("Model ", which.min(perf), " is lowest loss: ", min(perf), sep = "","\n")
Model 149 is lowest loss: 0.163925

> print(grid_search[which.min(perf), ])
    max_depth min_data_in_leaf min_sum_hessian_in_leaf feature_fraction bagging_fraction bagging_freq lambda_l1 lambda_l2 min_gain_to_split
149        10                2                    0.05              0.9              0.6            2       0.2       0.2               0.2
> 

```

**Comment by Laurae2** [comment]
@tobigithub Change the number of threads, and destroy + garbage collect the model everytime training finishes. It will lower the memory pressure a lot.

Something like this for the latter issue (destroy + garbage collect model):

```r
library(data.table)
library(lightgbm)

data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label, free_raw_data = FALSE)

data(agaricus.test, package = "lightgbm")
test <- agaricus.test
dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)

valids <- list(test = dtest)

grid_search <- expand.grid(max_depth = c(10,20,40,80),
                           min_data_in_leaf = c(1,2,4),
                           min_sum_hessian_in_leaf = c(0.05,0.1,0.2),
                           feature_fraction = c(0.8,0.9,0.95),
                           bagging_fraction = c(0.4,0.6),
                           bagging_freq = c(2,4),
                           lambda_l1 = c(0.2,0.4),
                           lambda_l2 = c(0.2,0.4),
                           min_gain_to_split = c(0.2,0.4))

perf <- numeric(nrow(grid_search))

for (i in 1:nrow(grid_search)) {
  model <- lgb.train(list(objective = "regression",
                          metric = "l2",
                          max_depth = grid_search[i, "max_depth"],
                          min_data_in_leaf = grid_search[i,"min_data_in_leaf"],
                          min_sum_hessian_in_leaf = grid_search[i, "min_sum_hessian_in_leaf"],
                          feature_fraction =  grid_search[i, "feature_fraction"],
                          bagging_fraction =  grid_search[i, "bagging_fraction"],
                          bagging_freq =  grid_search[i, "bagging_freq"],
                          lambda_l1 =  grid_search[i, "lambda_l1"],
                          lambda_l2 =  grid_search[i, "lambda_l2"],
                          min_gain_to_split =  grid_search[i, "min_gain_to_split"]),
                     dtrain,
                     2,
                     valids,
                     #min_data = 1,
                     num_leaves = 100,
                     learning_rate = 0.1,
                     early_stopping_rounds = 20)
  perf[i] <- min(rbindlist(model$record_evals$test$l2))
  gc(verbose = FALSE)
}
# grid_search
cat("Model ", which.min(perf), " is lowest loss: ", min(perf), sep = "","\n")
print(grid_search[which.min(perf), ])
```

**Comment by tarunparmar** [comment]
Hi @Laurae2 ,

I keep getting test auc of 1 on a kaggle dataset, am I doing something wrong?
```
[LightGBM] [Warning] Starting from the 2.1.2 version, default value for the "boost_from_average" parameter in "binary" objective is true.
This may cause significantly different results comparing to the previous versions of LightGBM.
Try to set boost_from_average=false, if your old models produce bad results
[LightGBM] [Info] Number of positive: 3120832, number of negative: 3124207
[LightGBM] [Info] Total Bins 7050
[LightGBM] [Info] Number of data: 6245039, number of used features: 73
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.499730 -> initscore=-0.001081
[LightGBM] [Info] Start training from score -0.001081
[1]:	test's auc:1 
[2]:	test's auc:1 
```

**Comment by tobigithub** [comment]
@tarunparmar 
yes, something is wrong with your code, probably overfitting.

**Comment by phdiims** [comment]
> @tobigithub Change the number of threads, and destroy + garbage collect the model everytime training finishes. It will lower the memory pressure a lot.
> 
> Something like this for the latter issue (destroy + garbage collect model):
> 
> ```r
> library(data.table)
> library(lightgbm)
> 
> data(agaricus.train, package = "lightgbm")
> train <- agaricus.train
> dtrain <- lgb.Dataset(train$data, label = train$label, free_raw_data = FALSE)
> 
> data(agaricus.test, package = "lightgbm")
> test <- agaricus.test
> dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)
> 
> valids <- list(test = dtest)
> 
> grid_search <- expand.grid(max_depth = c(10,20,40,80),
>                            min_data_in_leaf = c(1,2,4),
>                            min_sum_hessian_in_leaf = c(0.05,0.1,0.2),
>                            feature_fraction = c(0.8,0.9,0.95),
>                            bagging_fraction = c(0.4,0.6),
>                            bagging_freq = c(2,4),
>                            lambda_l1 = c(0.2,0.4),
>                            lambda_l2 = c(0.2,0.4),
>                            min_gain_to_split = c(0.2,0.4))
> 
> perf <- numeric(nrow(grid_search))
> 
> for (i in 1:nrow(grid_search)) {
>   model <- lgb.train(list(objective = "regression",
>                           metric = "l2",
>                           max_depth = grid_search[i, "max_depth"],
>                           min_data_in_leaf = grid_search[i,"min_data_in_leaf"],
>                           min_sum_hessian_in_leaf = grid_search[i, "min_sum_hessian_in_leaf"],
>                           feature_fraction =  grid_search[i, "feature_fraction"],
>                           bagging_fraction =  grid_search[i, "bagging_fraction"],
>                           bagging_freq =  grid_search[i, "bagging_freq"],
>                           lambda_l1 =  grid_search[i, "lambda_l1"],
>                           lambda_l2 =  grid_search[i, "lambda_l2"],
>                           min_gain_to_split =  grid_search[i, "min_gain_to_split"]),
>                      dtrain,
>                      2,
>                      valids,
>                      #min_data = 1,
>                      num_leaves = 100,
>                      learning_rate = 0.1,
>                      early_stopping_rounds = 20)
>   perf[i] <- min(rbindlist(model$record_evals$test$l2))
>   gc(verbose = FALSE)
> }
> # grid_search
> cat("Model ", which.min(perf), " is lowest loss: ", min(perf), sep = "","\n")
> print(grid_search[which.min(perf), ])
> ```

I am applying this on my own large dataset. The loop runs fine and the selected parameter values are returned. However, I also get the following error-

Error in data$update_params(params) : [LightGBM] [Fatal] Cannot change max_bin after constructed Dataset handle.
4.
data$update_params(params)
3.
(function (params = list(), data, nrounds = 10L, valids = list(), obj = NULL, eval = NULL, verbose = 1L, record = TRUE, eval_freq = 1L, init_model = NULL, colnames = NULL, categorical_feature = NULL, early_stopping_rounds = NULL, callbacks = list(), reset_data = FALSE, ...
2.
do.call(what = lgb.train, args = train_args)
1.
lightgbm(params = list(objective = "binary", metric = "binary_logloss", min_sum_hessian_in_leaf = A.lgb2.gridS[i, "min_sum_hessian_in_leaf"], feature_fraction = A.lgb2.gridS[i, "feature_fraction"], bagging_fraction = A.lgb2.gridS[i, "bagging_fraction"], bagging_freq = A.lgb2.gridS[i, "bagging_freq"], ...


Should I ignore this? Please help.

**Comment by tobigithub** [comment]
@phdiims the documentation for the  number of threads and other parameters for lgb.train() function is here

https://lightgbm.readthedocs.io/en/latest/R/reference/lgb.train.html
also this one with additional stuff and demos
https://www.rdocumentation.org/packages/Laurae/versions/0.0.0.9001/topics/lgbm.train

```
num_threads: Number of threads for LightGBM. For the best speed, set this to the number of real CPU cores, not the number of threads (most CPU using hyper-threading to generate 2 threads per CPU core)

#> [LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001144 seconds.
#> You can set `force_row_wise=true` to remove the overhead.
#> And if memory is not enough, you can set `force_col_wise=true`.
```

**Running your last script** I dont get any error, but I also can not get 100% CPU saturation, with _num_threads = -1_ or _num_threads = 8_ it still runs on one core, I think the method itself is too fast or too short, so spawning multiple threads within R may not work, I know it works in the executable (EXE).

But I also don't get any error with your code it runs fine. I am using R.3.3.1 plus I added two more lines.

````
                     num_threads = 8,
                     verbose = 5)
````

What it does, it converges very very slowly, you probably want to print the train and test curve (loss funtion) to follow in real-time. https://lightgbm.readthedocs.io/en/latest/_modules/lightgbm/plotting.html
````
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
[1]:    test's l2:0.202375 
[2]:    test's l2:0.164052 
`````
after that I just stopped it, because it went on my nerves, and it plots just the result, seems fine to me.

`````
> # grid_search
> cat("Model ", which.min(perf), " is lowest loss: ", min(perf), sep = "","\n")
Model 2828 is lowest loss: 0
> print(grid_search[which.min(perf), ])
     max_depth min_data_in_leaf min_sum_hessian_in_leaf feature_fraction
2828        80                2                     0.1              0.8
     bagging_fraction bagging_freq lambda_l1 lambda_l2 min_gain_to_split
2828              0.4            4       0.2       0.4               0.4
`````

From your error I only found two references, you might want to contact @guolinke or @Laurae2, they are the real masters of the code. I searched on the googles: [LightGBM] [Fatal] "Cannot change max_bin after constructed Dataset handle."

Max bin by feature #2190
https://github.com/microsoft/LightGBM/pull/2190/files/cf4707859318f13888aef8c2850b4e48d1d72f03


Lightgbm with GridSearch [Starter] (no error)
https://www.kaggle.com/theoverfitter3139/lightgbm-with-gridsearch-starter-1

Lightgbm with GridSearch [Starter] (with  error)
https://www.kaggle.com/theoverfitter3139/lightgbm-with-gridsearch-starter-1?scriptVersionId=11902060


So thats the best I can do,  check the Kaggle example, there are many more parameters that can be used,
also run your code on the Kaggle cloud,  to see if it works there, hanging out on Kaggle really can help.

On a side note I stopped using R (because of memory and threading issues) and the DLL and version dependency hell and switched to Python for machine learning.
